interfaces/builtin/pkcs11_test.go

```diff
@@ -0,0 +1,277 @@
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-1-slot"), testutil.Contains,
+		`/usr/bin/pkcs11-tool ixr,`)
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-1-slot"), testutil.Contains,
+		`"/{,var/}run/p11-kit/pkcs11-optee-slot-0" rw,`)
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-1-slot"), Not(testutil.Contains),
+		`"/{,var/}run/p11-kit/pkcs11-optee-slot-1" rw,`)
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-1-slot"), Not(testutil.Contains),
+		`"/{,var/}run/p11-kit/pkcs11-optee-slot-2" rw,`)
+
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-2-slots"), testutil.Contains,
+		`"/{,var/}run/p11-kit/pkcs11-optee-slot-0" rw,`)
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-2-slots"), testutil.Contains,
+		`"/{,var/}run/p11-kit/pkcs11-optee-slot-1" rw,`)
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-2-slots"), Not(testutil.Contains),
+		`"/{,var/}run/p11-kit/pkcs11-optee-slot-2" rw,`)
+
+	err = apparmorSpec.AddConnectedPlug(s.iface, s.testPlug2, s.testSlot2)
+	c.Assert(err, IsNil)
+	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.consumer.app-accessing-1-slot", "snap.consumer.app-accessing-2-slots", "snap.consumer.app-accessing-3rd-slot"})
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-3rd-slot"), Not(testutil.Contains),
+		`"/{,var/}run/p11-kit/pkcs11-optee-slot-0" rw,`)
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-3rd-slot"), Not(testutil.Contains),
+		`"/{,var/}run/p11-kit/pkcs11-optee-slot-1" rw,`)
+	c.Assert(apparmorSpec.SnippetForTag("snap.consumer.app-accessing-3rd-slot"), testutil.Contains,
+		`"/{,var/}run/p11-kit/pkcs11-optee-slot-2" rw,`)
+}
+
+func (s *Pkcs11InterfaceSuite) TestSanitizeGadgetSnapSlots(c *C) {
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.testSlot0Info), IsNil)
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.testSlot1Info), IsNil)
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.testSlot2Info), IsNil)
+}
+
+func (s *Pkcs11InterfaceSuite) TestSanitizeBadGadgetSnapSlots(c *C) {
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.testBadSlot0Info), ErrorMatches, "slot \"pkcs11-bad-optee-slot-0\", a unix socket has to be in /run/p11-kit directory")
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.testBadSlot1Info), ErrorMatches, "slot \"pkcs11-bad-optee-slot-1\", a unix socket name has to start with 'pkcs11-'")
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.testBadSlot2Info), ErrorMatches, "slot \"pkcs11-bad-optee-slot-2\", a unix socket name has to start with 'pkcs11-'")
+	c.Assert(interfaces.BeforePrepareSlot(s.iface, s.testBadSlot3Info), ErrorMatches, "pkcs11: \"pkcs11-socket\" attribute must be a string, not 22")
+	err := errors.New("pkcs11 interface socket path is invalid: \"/run/p11-kit/pkcs11-optee-slot-*\" contains a reserved apparmor char from ?*[]{}^\"\x00")
```

Comment:
I think this line is wrong :smile: I think you meant to create the error by calling BeforePrepareSlot on `testBadSlot4Info` ?

Reply:
seems so indeed